src/cls-ah.js

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const asyncHook = require('async_hooks');
+
+const wrappedSymbol = Symbol('context_wrapped');
+const contexts = {};
+var current = new Context();
+
+const hook = asyncHook.createHook({init, before, after, destroy});
+hook.enable();
+
+// Core Zone Api
+
+function Context() {
+  this.properties = {};
+}
+
+Context.prototype.set = function(key, val) {
+  this.properties[key] = val;
+};
+
+Context.prototype.get = function(key) {
+  return this.properties[key];
+};
+
+Context.prototype.wrap = function(cb) {
+  if (cb[wrappedSymbol]) {
+    return cb;
+  }
+  const oldContext = current;
+  const thisContext = this;
+  const wrapped = cb.length === 0 ? function() {
+    current = thisContext;
+    const res = cb.apply(this, arguments);
+    current = oldContext;
+    return res;
+  } : function(a) {
+    current = thisContext;
+    const res = cb.apply(this, arguments);
+    current = oldContext;
+    return res;
+  };
+  wrapped[wrappedSymbol] = true;
+  return wrapped;
+};
+
+Context.prototype.run = function(cb, applyThis, applyArgs) {
+  const oldContext = current;
+  current = new Context();
```

Comment:
Question: there are three contexts objects in this function: `current`, `new Context` and `this`. I believe `this === current` at the start of the function?